### ISSUE TITLE
MessageProcessingAbortedException looks like a smell

### ISSUE THREAD

**Issue by tmasternak** [Particular/NServiceBus#3044]
We have a special type of exception which is MessageProcessingAbortedException. Currently it is used to indicate that the in the pipeline failed in a 'controlled-way' e.g. FLR, SLR or Faults behaviors have caught the exception and handled it. 
To the user of a pipeline it means that processing has been aborted but should not be counted as an error which currently results in: 
 # Transaction (if existing) being roll-backed. 
 # No error statistics update in circuit breaker.
I think that it current design can be a problem in the future for someone writing new transport. It's not clear that such exception exists and I do not thing that it's exceptional situation we are talking about.

**Comment by SeanFeldman** [comment]
To add to this @tmasternak, I **had no idea** that it has to be implemented while working on Particular supported transport - ASB.

**Comment by johnsimons** [comment]
A huge :+1: from me, this is just a code smell

**Comment by weralabaj** [comment]
@tmasternak What's the proposed solution? Keep it with better name and doco or refactor to get rid of it?

**Comment by andreasohlund** [comment]
>A huge :+1: from me, this is just a code smell

There was ideas to have the core catch it and have  https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Transports/IPushMessages.cs#L14 return `Task<bool>` instead. This would lead to the transports doing

```
if(!await pipe(...))
{
    //roll message back to the queue
}

```

What that be a better api?

**Comment by tmasternak** [comment]
@weralabaj I do not have a solution here. I guess that #3220 and proper documentation will be a move in a good way.

**Comment by johnsimons** [comment]
Not sure @andreasohlund, would you be able to show us a bit more code ?
From what I see in your snippet, it seems still very up to the transport implementor to figure all of this out, which is IMO quite error prone.

**Comment by tmasternak** [comment]
@andreasohlund @johnsimons  To add to that I think that bool is too little information. When the pipline does not handle the message (returns false) we still do not know if we should count it for circuit breaker or not.

**Comment by johnsimons** [comment]
Well, we could return not a bool, but something more informational.
I will have a beter look tomorrow and see what I can come up with.

**Comment by tmasternak** [comment]
This will be solved when we finish #3242. Closing
